My Jetpack: Introduce main section components

leogermani opened the pull request:
<!--- Provide a general summary of your changes in the Title above -->
<!-- Would you like this feature to be tested by Beta testers?
Please add testing instructions to projects/plugins/jetpack/to-test.md in a new commit as part of your PR. -->

This PR proposes two new fundamental layout components to our Design system and applies them to the My Jetpack Screen.

## Jetpack Admin Page

This component is basically the main container for any admin page of any plugin. It adds the Header and the Footer.

## Jetpack Admin Section

This component is the wrapper for each section (or each row) of the admin screen. The propose is that anything in the admin page should be added in one of these and this will ensure all sections have the same spacing and behave the same way in every break point.

### Options

**bgColor:** There are 2 possible background colors: white and grey

Here is how they look and behave:

![Captura de Tela 2021-11-26 às 15 22 48](https://user-images.githubusercontent.com/971483/143620967-24ca0fdd-37e7-4f6b-922c-df3656cef861.png)

![Captura de Tela 2021-11-26 às 15 33 20](https://user-images.githubusercontent.com/971483/143620980-873a966e-8eb3-46dd-b9a6-c257ef3cad08.png)


#### Changes proposed in this Pull Request:
<!--- Explain what functional changes your PR includes -->
* Adds JetpackAdminPage and JetpackAdminSection components.

#### Jetpack product discussion
<!-- If you're an Automattician, include a shortlink to the p2 discussion with Jetpack Product here. -->
<!-- Make sure any changes to existing products have been discussed and agreed upon -->
N/A

#### Does this pull request change what data or activity we track or use?
<!--- If so, please add the "[Status] Needs Privacy Updates" label and explain what changes there are. -->
<!--- Check existing Jetpack support documents for a preview of the information we need. -->

#### Testing instructions:
<!-- If you were reviewing this PR, how would you like the instructions to be presented? -->
<!-- Please include detailed testing steps, explaining how to test your change. -->
<!-- Bear in mind that context you working on is not obvious for everyone.  -->
<!-- Adding "simple" configuration steps will help reviewers to get to your PR as quickly as possible. -->
<!-- "Before / After" screenshots can also be very helpful when the change is visual. -->

On Storybook:

* Checkout this branch
* `jetpack build js-packages/components`
* `cd projects/js-packages/storybook`
* `pnpm run storybook:dev`
* It will open your browser
* Go to PlayGround > Admin Page and Playground > Admin Section

On a live site:

* Check out this branch or launch a JN site in this branch
* Go to Settings > Jetpack Constants and enable `JETPACK_ENABLE_MY_JETPACK`
* Go to Jetpack > My Jetpack

Question to design:

Are the options offered by these wrapper elements enough to cover all our scenarios so far? Are the spacings right? Do we need any additional option?

Looking at some of the designs, it seems that the top margin of the row with a grey background is slightly smaller than the other section. Any suggestions on how to treat this? Instead of a `bgColor` option should we have something like `Hero` or `highlight` option that would add the grey background and decrease the top margin?

leogermani committed:
create AdminSection component

leogermani committed:
use adminSection in my jetpack

leogermani committed:
[not verified] Merge branch 'master' into add/my-jetpack-sections

leogermani committed:
created AdminPage and refactor AdminSection

leogermani committed:
internal deps

leogermani committed:
changelog

renatoagds commented on projects/js-packages/components/components/admin-section/index.jsx:
```diff
@@ -0,0 +1,44 @@
+	let mainClassName = 'jp-admin-section';
+	if ( bgColor === 'grey' ) {
+		mainClassName += ' jp-admin-section--bg-grey';
+	}
```
You could start to use [classnames](https://www.npmjs.com/package/classnames) utility here.

```
const mainClassName = classNames(
  'jp-admin-section',
  { 'jp-admin-section--bg-grey': bgColor === 'grey' }
);
```
Looks like we already have it on the project: 

https://github.com/Automattic/jetpack/search?q=classnames

leogermani replied:
thanks! Will look into it

leogermani replied:
Done in https://github.com/Automattic/jetpack/pull/21826/commits/70c68739a422181451b358854e424b402ca4563f

renatoagds commented on projects/packages/my-jetpack/_inc/components/my-jetpack-screen/style.scss:
```diff
@@ -0,0 +1 @@
+@import '~@automattic/jetpack-base-styles/style';
```
Just curious about why we need to put in all places this impor, it's not injected globally?

leogermani replied:
I don't know, are they?

Here again, I'm hoping we refactor all this CSS very soon.

By very soon I mean before someone starts to use and customize these elements using their classes

sergeymitr replied:
It's my understanding that we're moving away from importing the [`style.scss`](https://github.com/Automattic/jetpack/blob/master/projects/js-packages/base-styles/style.scss) anyway, as those styles get applied globally and affect non-RNA parts of the page.

So I think it's ok to keep importing this styles file as a temporary workaround as is until we get the styling roadblock resolved.

renatoagds replied:
Agreed @sergeymitr

leogermani commented on projects/js-packages/components/components/admin-page/index.jsx:
```diff
@@ -0,0 +1,63 @@
+						</div>
+					</div>
+				</div>
+			</div>
```
That was the idea at first, but later I saw little benefit in doing that.

Especially because we are about to refactor the whole CSS structure, and this would change anyway. I'm only concerned with the public API right now (the JetpackAdminPage element and its props).

But if I was to do it, I'd actually create a Header and a Footer internal components, and the component would be basically:

```
<Header />
{ children }
<Footer />
```

leogermani replied:
Done in https://github.com/Automattic/jetpack/pull/21826/commits/2a1dfa7d64730974c7991458fff8575d263d8cf2

leogermani committed:
break adminPage in sub components

leogermani committed:
use classnames in adminSection

leogermani committed:
[not verified] Merge branch 'master' into add/my-jetpack-sections

leogermani committed:
improve readability

renatoagds commented on projects/js-packages/components/components/admin-section/style.scss:
```diff
@@ -0,0 +1,18 @@
+@import '~@automattic/jetpack-base-styles/style';
+
+// This should go to base styles?
```
That's a good question. Any details why we're applying these styles?

sergeymitr replied:
I believe changes to `#wpcontent` and `#wpwrap` styling should not go into any RNA packages.
I'd rather we wouldn't edit them at all, but if we do, we should put that in the Composer My Jetpack package, not in the RNA.
Otherwise somebody will get their whole WP-admin styling modified by simply importing an RNA component. I don't think we should allow that to happen.

renatoagds replied:
@leogermani can you explain more for me what these changes do at wp structure? (just to know more about it).

renatoagds replied:
Ok. Saw your comment here: https://github.com/Automattic/jetpack/pull/21826#issuecomment-981912803

sergeymitr commented on projects/js-packages/components/components/admin-page/index.jsx:
```diff
@@ -0,0 +1,47 @@
+ * @param {object} props - Component properties.
+ * @returns {React.Component} JetpackAdminPage component.
+ */
+const JetpackAdminPage = props => {
```
@leogermani, what do you think about moving `JetpackAdminPage`, `JetpackAdminPageHeader`, and `JetpackAdminPageFooter` components to the RNA Components package?
Since we're aiming at same base design for all consumers, should we reuse them for Jetpack plugins as well?

keoshi commented:
This is cool, thanks for getting it started @leogermani !

> Are the options offered by these wrapper elements enough to cover all our scenarios so far?

Yes, these should be enough for the foreseeable future in RNA.

> Looking at some of the designs, it seems that the top margin of the row with a grey background is slightly smaller than the other section. Any suggestions on how to treat this? Instead of a bgColor option should we have something like Hero or highlight option that would add the grey background and decrease the top margin?

Yes, the grey block should have a top padding of `48px` (vs the highlighted `64px`). I really, really like the option of having a prop that defines it being an `hero` or not. Then we would treat all blocks equally, except for this one which changes the background color and top padding automatically.

sergeymitr commented on projects/js-packages/components/components/admin-section/stories/index.jsx:
```diff
@@ -0,0 +1,29 @@
+// Export additional stories using pre-defined values
+const Template = args => (
+	<JetpackAdminPage>
+		<JetpackAdminSection { ...args }>
+			Sample Section 1 - Controls will apply to this section
+		</JetpackAdminSection>
+		<JetpackAdminSection>Sample Section 2</JetpackAdminSection>
```
Should we split these two sections into two separate stories, like we did for the [`PricingCard`](https://github.com/Automattic/jetpack/blob/master/projects/js-packages/components/components/pricing-card/stories/index.jsx)?

leogermani replied:
Yes, I could create additional stories here.

I just wanted to have a story with more than one Section to help us visualize.

sergeymitr replied:
That makes a lot of sense, actually, I didn't think about it that way.

sergeymitr reviewed (commented):
Hi @leogermani.
Nicely done!
I left a couple of comments, please let me know what you think.

leogermani commented:
> I really, really like the option of having a prop that defines it being an hero or not. Then we would treat all blocks equally, except for this one which changes the background color and top padding automatically.

Thanks for the reply @keoshi . I also prefer this approach rather than having a `bgColor` prop. But even better, I think I'll make them two different components. `JetpackSection` and `JetpackHero`. What do you think?

leogermani commented:
> I believe changes to #wpcontent and #wpwrap styling should not go into any RNA packages.
I'd rather we wouldn't edit them at all, but if we do, we should put that in the Composer My Jetpack package, not in the RNA.
Otherwise somebody will get their whole WP-admin styling modified by simply importing an RNA component. I don't think we should allow that to happen.

@sergeymitr , I got this from the Backup plugin. They do it to make sure there's no margin between the left menu and the "Hero" grey background. I'll give it a try and see if I can find another solution for that without messing with these global styles

leogermani committed:
create JetpackAdminSectionHero component

leogermani committed:
remove CSS for wp-admin core elements

leogermani commented:
@sergeymitr , @renatoagds , as I chatted with @keoshi in the comments above, we agreed that the best interface for developers is to have an `hero` option, instead of `bgColor` and `topPadding` options. I went ahead and created two independent components for this.

Please have a look at https://github.com/Automattic/jetpack/pull/21826/commits/67cf1fc34b9810d5b2662adb7330cde2df09a135 and see what you think. 

At first, I thought of creating a third base component, that would not be exported, but would serve as a base for both components, but in the end I decided to duplicate some code. The reason was that I would have to create a component with props and I didn't think it would escalate very well if we wanted to create other kinds of Sections. Also, since we are moving to scoped CSS, I couldn't simply import the base component and modify it via CSS...

@keoshi if you could also have a look at how it looks. I'm looking for confirmation on the left margin now. Its size and how it should behave in narrow screens. Thanks!

leogermani commented:
@sergeymitr , @renatoagds, @keoshi 

On another comment, I removed some CSS that were being applied to root wp-admin components. 

I added a left margin of `-20px` to the main AdminPage element, to neutralize the padding we have in `#wpcontent`. I think it works fine.

But we also had a white background for the whole admin area. I don't think we can change that easily. But also I don't think we should.

This will only make a difference when we don't have enough content to fill the vertical size of the screen. Here it how it looks:

![Captura de Tela 2021-11-29 às 18 44 03](https://user-images.githubusercontent.com/971483/143948975-6d9ba4f4-3164-498e-9f58-babcee5e2a68.png)

We'll have that grey background under the content. Do you think this is a problem?

renatoagds commented:
@leogermani I have some thoughts about the component duplication, since it looks we're not making use of react `props` that was designed exactly for this purpose: ability to change the component.

> The reason was that I would have to create a component with props and I didn't think it would escalate very well if we wanted to create other kinds of Sections

Maybe isn't the case to create the second one when we hit the complex case? I'm thinking that we could have the `AdminSection` with `hero` prop, and if something really different appears, this one will be a new component. Makes sense?

> But we also had a white background for the whole admin area

We could move the footer to be fixed at the bottom, not? This will increase the height and use our BG.

leogermani committed:
[not verified] Merge branch 'master' into add/my-jetpack-sections

leogermani committed:
remove Jetpack prefix from components names

leogermani committed:
update internal deps

leogermani commented:
@renatoagds @sergeymitr , I just updated the PR, removed the `Jetpack` prefix from the components name. Leaving it as 2 separate components. Would appreciate a final check, thanks!

renatoagds reviewed (approved):
LGTM 🚀

leogermani committed:
Merge branch 'master' into add/my-jetpack-sections

fgiannar reviewed (approved):
Thanks for working on this, Leo!

Left only a minor comment for future reference though :)

leogermani commented:
r235966-wpcom
